### ISSUE TITLE
How is S4 used in the implementation?  (I may be mistaken, but I think s4 is used differently in this implementation than the paper description )

### ISSUE THREAD

**Issue by zmy1116** [md-mohaiminul/TranS4mer#4]
Hello,

Following the previous question on  finetuning code, I start to read through the full codebase to better understand your work. 

I have some questions regarding the code because I'm a bit confused about how S4 is used in the implementation.

Based on my understanding of the paper:
- For every shot, we sample `K` frames, for each frames we run ViT and obtain `P` patche tokens, in total `PxK` tokens
- Given we have `N` shots,  we have then `N x PxK` tokens,  we flatten them out and run S4 model on the long sequence.

What I see in the code is following:

- we process each frame with VIT,  but we use `S4` module instead of normal self attention 
https://github.com/md-mohaiminul/TranS4mer/blob/98de379770e0d4034280abb9923928c103377286/trans4mer/pretrain/cfg/default.yaml#L88-L91
https://github.com/md-mohaiminul/TranS4mer/blob/98de379770e0d4034280abb9923928c103377286/trans4mer/model/timm/models/vision_transformer.py#L248-L253

- To handle `N` shots, we are using the same `trn`  proposed in bassl,  a BERT encoder. 

https://github.com/md-mohaiminul/TranS4mer/blob/98de379770e0d4034280abb9923928c103377286/trans4mer/pretrain/pretrain_wrapper.py#L147-L177

So we generate first generate shot level representation with `extract_shot_representation`,   and then run the `self.loss`, which is the `BaSSLLoss`

The `BaSSLLoss` use `model.crn` to aggregate information across multiple shots,  the `crn` is the BERT encoder used in BASSL, not an S4 module. 

I'm following the procedure defined by `default.yaml` config.  And it appears that the implemented procedure is significantly different than what the paper proposed.   

I'm not sure if I'm reading the right code portion or I misunderstood the paper.    I appreciate greatly if you can provide some guidance. 

Thank you

**Comment by zmy1116** [comment]
Hello,

I understand that the `s4` module actually go through all shots.
- we collect the kth frame (out of K) per shot 
- given s shots,  we pass all s frames into the shot encoder 
- we get s x k features at the end,  we then average over k  to get per shot representation 

Thanks, I'm closing this issue
